Sources/NIO/Bootstrap.swift

```diff
@@ -398,10 +364,21 @@ internal protocol NIOTCPClientBootstrap {
 /// ```
 ///
 /// The connected `SocketChannel` will operate on `ByteBuffer` as inbound and on `IOData` as outbound messages.
-public final class ClientBootstrap: NIOTCPClientBootstrap {
-
+public final class ClientBootstrap: NIOClientTCPBootstrapProtocol {
     private let group: EventLoopGroup
-    private var channelInitializer: Optional<ChannelInitializerCallback>
+    private var protocolHandlers: Optional<() -> [ChannelHandler]>
+    private var _channelInitializer: ChannelInitializerCallback
+    private var channelInitializer: ChannelInitializerCallback {
+        if let protocolHandlers = self.protocolHandlers {
+            return { channel in
+                self._channelInitializer(channel).flatMap {
+                    channel.pipeline.addHandlers(protocolHandlers(), position: .first)
```

Comment:
What's the reasoning for doing initializer followed by protocol handlers? I worry about channel handlers issuing writes in their `handlerAdded`.

Reply:
@Lukasa good Q but the reasoning is that we can't reverse it because if the user does `addHandler(..., position: .first)` in the `channelInitializer`, then that would break TLS. Protocol handlers need to go at the very front... Writes from `handlerAdded` are indeed a problem, writes from `channelActive` are not

Reply:
I figured. Ok, good to know, my approval from earlier stands.